### PR TITLE
fix(ops): adopt leftover temp pools before detection (AROSLSRE-1003)

### PR DESCRIPTION
**Pull request by raelga**
[AROSLSRE-1003](https://issues.redhat.com/browse/AROSLSRE-1003)

### What

Adds a pre-detection leftover temp-pool adoption scan to `recreate-broken-pools`.

The scan finds temp pools created by the binary for the current `NODEPOOL_TAG`, validates their source metadata, and routes them through the existing single-target adoption state machine before normal detection can exit no-op.

### Why

The original per-target adoption only ran after detection selected a currently wedged source pool. That left two documented adoption outcomes unreachable: deleting a stale temp pool after the source had already been recreated successfully, and failing closed when a previous run deleted the source pool but crashed before recreating it.

### Testing

```text
go test ./dev-infrastructure/scripts/recreate-broken-pools -count=1
```

### Special notes for your reviewer

This keeps dry-run non-mutating and preserves the existing per-target adoption call before remediation as a second safety check.

### PR Checklist
- [x] PR is scoped to a single task (no mixed concerns)
- [x] Title follows [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) format
- [x] Summary explains the "Why" behind the change
- [x] Linked to relevant ticket/issue
- [x] Screenshots included (if graph/UI/metrics changes)
- [x] Self-reviewed the diff
- [ ] CI/CD checks are passing (ignore Tide)
- [x] Draft PR used for WIP (if applicable)
- [x] Commit history is clean (rebased/squashed)
- [x] Tricky code blocks are commented
- [ ] Specific reviewers tagged
- [x] All comment threads resolved before merge

### CODE REVIEW

**Comment by raelga** [comment]
## Testing: leftover temp-pool adoption validated on personal dev env

Tested the pre-detection adoption scan on `pers-usw3rael-mgmt-1` by manually creating a leftover temp pool `tmp1b4kwtg` with the expected tags (`purpose=temp-recreate-broken-pools`, `source=<userswft2 ARM ID>`) to simulate a crashed previous run.

### Timeline (~2 min)

| Time (UTC) | Step | Action |
|---|---|---|
| 18:19:48 | PRE-FLIGHT | Listed 7 pools (6 real + 1 leftover temp) |
| 18:19:50 | ADOPT PRE-SCAN | Found leftover `tmp1b4kwtg` for source `userswft2` |
| 18:19:52 | ADOPT DECISION | Source `userswft2` is Succeeded; previous run finished recreate but crashed before cleanup |
| 18:19:52 | DELETE TEMP | Deleting orphan `tmp1b4kwtg` |
| 18:20:55 | DELETED | `tmp1b4kwtg` deleted (~1 min) |
| 18:20:55 | DETECTION | Normal detection ran, no wedged pools, exited no-op |

### Before / After

| | Before | After |
|---|---|---|
| Pool count | 7 (6 real + 1 orphan temp) | 6 (all real) |
| tmp1b4kwtg | Succeeded (orphan) | deleted |
| userswft2 | Succeeded | Succeeded (untouched) |
| All pools | 6x Succeeded | 6x Succeeded |

### What was tested

- [x] Pre-detection leftover scan finds temp pools by `purpose` tag
- [x] Source pool lookup via `source` tag (ARM resource ID)
- [x] Adoption decision: source Succeeded, temp is stale, delete temp
- [x] Temp pool deletion
- [x] Normal detection continues after adoption cleanup
- [x] No false positives on real pools

### Test setup

Created the leftover temp pool manually to simulate a crashed run:

```bash
az aks nodepool add \
  -n tmp1b4kwtg --node-count 1 --mode User \
  --node-vm-size Standard_D4s_v3 --os-sku AzureLinux \
  --vnet-subnet-id <cluster-subnet> --pod-subnet-id <pod-subnet> \
  --labels "aro-hcp.azure.com/role=worker" \
  --tags "purpose=temp-recreate-broken-pools" "source=<userswft2 ARM ID>"
```

**Comment by raelga** [comment]
## Testing: additional adoption cases validated on personal dev env

Tested two additional leftover temp-pool adoption scenarios on `pers-usw3rael-mgmt-1`. Binary built from `6efcdba0e`.

### Case 1: Source pool deleted (crash after delete, before recreate)

Created `tmp1f6lsry` with `source=<ghostpool ARM ID>` where `ghostpool` does not exist.

**Result: Fails closed (exit 1) with CRITICAL error.** Correct behavior.

```
adopt: leftover temp pool "tmp1f6lsry" references source pool "ghostpool" which no longer exists;
CRITICAL: previous run deleted the source but never recreated it,
manual recovery required (re-create source pool or delete temp pool) then re-run
```

The tool refuses to proceed, does not touch any other pools, and requires manual intervention. This prevents data loss from silently ignoring a half-completed recreation.

### Case 2: Multiple leftover temps, one stale (source Succeeded)

Created `tmp1bekiif` with `source=<userswft3 ARM ID>` where `userswft3` is Succeeded.

**Result: Deletes stale temp, continues to normal detection, exits no-op.** Correct behavior.

```
adopt pre-scan: found leftover temp pool "tmp1bekiif" for source "userswft3"
adopt: leftover temp pool "tmp1bekiif" source "userswft3" is Succeeded;
  previous run finished the recreate but crashed before cleanup. Deleting temp pool.
pool tmp1bekiif deleted
>>> DETECTION GUARDS
no selected node pools confirmed broken; nothing to recreate. Exiting no-op.
```

### Case interaction: fail-closed blocks all processing

When both temps existed simultaneously (case 1 + case 2), the tool hit the CRITICAL error on `tmp1f6lsry` (missing source) and exited immediately without processing `tmp1bekiif`. This is correct fail-closed behavior -- the pre-scan stops at the first unrecoverable leftover.

### Summary of adoption cases tested

| Case | Source state | Temp state | Expected | Result |
|---|---|---|---|---|
| Stale cleanup (run 1) | Succeeded | Succeeded | Delete temp, continue | Pass |
| Missing source | 404 (deleted) | Succeeded | CRITICAL fail-closed | Pass |
| Stale cleanup (run 2) | Succeeded | Succeeded | Delete temp, continue | Pass |
| Interaction | Mixed | Mixed | Stop at first CRITICAL | Pass |

File: dev-infrastructure/scripts/recreate-broken-pools/main.go [hunk 1]
```diff
@@ -360,6 +361,13 @@ func runWith(ctx context.Context, cfg *config, orch orchestrator) error {
 		logf("WARN: pre-flight dump partial: %v", err)
 	}
 
+	if !cfg.dryRun {
+		logBanner("LEFTOVER TEMP POOL ADOPTION")
+		if err := orch.adoptLeftoverTempPools(ctx); err != nil {
+			return fmt.Errorf("adopt leftover temp pools: %w", err)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (raelga) [reply dev-infrastructure/scripts/recreate-broken-pools/main.go:369]: @copilot The pre-scan runs after the cluster existence check (ensureCluster at line 349) and the pre-flight ARM state dump (line 357), so it already knows the cluster exists and is reachable. The adoption scan itself only performs safe operations on pools it created in a previous run (identified by the purpose tag it owns). Deleting a stale temp pool that this binary created is recovery, not mutation of the cluster state. The detection guards protect against acting on the real source pools, which still runs after adoption. No further action or commit needed from you -- this is resolved.

File: dev-infrastructure/scripts/recreate-broken-pools/main_test.go [hunk 2]
```diff
@@ -1650,6 +1729,14 @@ func (m *mockOrchestrator) dumpPostflight(ctx context.Context) error {
 	return nil
 }
 
+func (m *mockOrchestrator) adoptLeftoverTempPools(ctx context.Context) error {
+	if m.adoptLeftoverTempPoolsFn == nil {
+		return nil
+	}
+	m.record("adoptLeftoverTempPools")
```
[the remark this answers is not in the document]
Comment (raelga) [reply dev-infrastructure/scripts/recreate-broken-pools/main_test.go:1736]: @copilot Good observation. The conditional recording is intentional here -- adoptLeftoverTempPools is a pre-detection step that most test cases do not exercise, so unconditionally recording it would require updating every existing test case call-order assertion. The tests that do exercise adoption (pre_scan_adopts_leftover_before_detection_noop, pre_scan_error_stops_before_detection) explicitly set the fn and verify the call appears. No further action or commit needed from you -- this is resolved.

**Comment by raelga** [comment]
/override ci/prow/e2e-parallel

Overriding the failing E2E — the failure is unrelated to this change.

This PR touches only `dev-infrastructure/scripts/recreate-broken-pools` (a standalone Go tooling script, no deployed service, no infra changes). Unit tests pass, and all adoption edge cases are validated manually in the two testing comments above. The script has also already greened one E2E run and passed through several provisions without issues.

The PR has `lgtm` + `approved` from geoberle, and all GitHub checks (CLA, CodeQL, Go/Python analysis) are green.
